Mathlib/Topology/Algebra/Nonarchimedean/TotallyDisconnected.lean

```diff
@@ -0,0 +1,123 @@
+/-
+Copyright (c) 2024 Jou Glasheen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jou Glasheen, Kevin Buzzard
+-/
+
+import Mathlib.Topology.Algebra.Nonarchimedean.Basic
+
+/-!
+# Total disconnectedness of nonarchimedean groups
+
+In this file, we prove that a nonarchimedean group is a totally disconnected topological space.
+
+## Main results
+
+- `NonarchimedeanGroup.instTotallyDisconnectedSpace`: A nonarchimedean group is a
+                                                      totally disconnected topological space.
+
+## Notation
+
+ - `G` : Is a nonarchimedean group.
+ - `V` : Is an open subgroup which is a neighbourhood of the identity in `G`.
+
+## References
+
+See Proposition 2.3.9 and Problem 63 in :
+Gouvêa, F. Q. (2020) p-adic Numbers An Introduction. 3rd edition.
+  Cham, Springer International Publishing.
+-/
+
+namespace Set
+
+lemma mem_unique_to_singleton {X : Type*} {U : Set X} {x : X} (hx : x ∈ U)
+    (h : ∀ y : X, y ∈ U → y = x) : U = {x} := by
+  ext
+  constructor <;> simp_all
+
+end Set
+
+open Pointwise TopologicalSpace
+
+namespace NonarchimedeanGroup.auxiliary
+
+variable (G : Type*) [TopologicalSpace G] [Group G] [NonarchimedeanGroup G] [T2Space G]
+
+@[to_additive]
+lemma open_subgroup_separating
+    (t : G) (ht : t ≠ 1) : ∃ (A : Opens G) (V : OpenSubgroup G),
+    t ∈ A ∧ 1 ∈ V ∧ Disjoint (A : Set G) V := by
+  rcases (t2_separation ht) with ⟨A, B, opena, openb, diff, one, disj⟩
+  obtain ⟨V, hV⟩ := NonarchimedeanGroup.is_nonarchimedean B (IsOpen.mem_nhds openb one)
+  exact ⟨⟨A, opena⟩, V, diff, one_mem V,
+    fun _ ↦ by apply Set.disjoint_of_subset_right hV disj⟩
+
+end NonarchimedeanGroup.auxiliary
+
+namespace NonarchimedeanGroup
+open NonarchimedeanGroup.auxiliary
+
+variable (G : Type*) [TopologicalSpace G] [Group G]
+
+@[to_additive]
+lemma non_empty_intersection_compl_coset (x y : G) (U : Set G) (hx : x ∈ U)
+    (A : Opens G) (quot : (y⁻¹ * x) ∈ A ) (V : OpenSubgroup G) (dva : Disjoint (V : Set G) A) :
+    (U ∩ ((y • (V : Set G))ᶜ)).Nonempty := by
+  simp_rw [Set.inter_nonempty, Set.mem_compl_iff]
+  use x, hx
+  intro con
+  rw [mem_leftCoset_iff y] at con
+  have mem : (y⁻¹ * x) ∈ (V : Set G) ∩ A := Set.mem_inter con quot
+  have subempty : (V : Set G) ∩ A = ∅ := Disjoint.inter_eq dva
+  rw [subempty] at mem
+  simp at mem
+
+@[to_additive]
+lemma non_empty_intersection_coset (y : G) (U : Set G) (hy :  y ∈ U)
```

Comment:
To move these lemmas to the file for topological groups, should I just put them in a `TopologicalGroup` namespace? 

Thank you very much for reviewing my work!

Reply:
> To move these lemmas to the file for topological groups, should I just put them in a `TopologicalGroup` namespace?
> 
> Thank you very much for reviewing my work!

Yes, they should be in the `TopologicalGroup` namespace. If you think this lemma won't be used in the future, you can also choose not to move it to the `TopologicalGroup` file and instead change it to a `private lemma`.

Reply:
> > To move these lemmas to the file for topological groups, should I just put them in a `TopologicalGroup` namespace?
> > Thank you very much for reviewing my work!
> 
> Yes, they should be in the `TopologicalGroup` namespace. If you think this lemma won't be used in the future, you can also choose not to move it to the `TopologicalGroup` file and instead change it to a `private lemma`.

So do you want them to be `private lemma`?

Reply:
If I change them to private lemma, the `@[to_additive]` tag does not work. Do you have any suggestions?

Reply:
This may be a bug: after setting it as a private lemma, `@[to_additive]` does not transform the Group in the namespace to AddGroup. This causes the lemma name to remain the same after the `@[to_additive]` operation, resulting in an error. I think you can remove the namespace and change the name to something like `non_empty_intersection_group_coset`.